📖 doc: add local e2e test execution documentation (#4626)
kubernetes-sigs/cluster-api: docs/book/src/developer/testing.md

See [e2e development] for more information on developing e2e tests for CAPI and external providers.

## Running the end-to-end tests locally

> Usually, the e2e tests are executed by Prow either pre-submit (on PRs) or periodically on certain branches

Reviewer: minor nit
```suggestion
Usually the e2e tests are executed by Prow, either pre-submit (on PRs) or periodically on certain branches
```

(e.g. the default branch). Those jobs are defined in the kubernetes/test-infra repository in [config/jobs/kubernetes-sigs/cluster-api](https://github.com/kubernetes/test-infra/tree/master/config/jobs/kubernetes-sigs/cluster-api).
For development and debugging those tests can also be executed locally.

### Prerequisites

> 

Reviewer: does the user need anything beyond docker and the make/build commands?

Author: I think it depends on which exact test is executed and in which way it is executed. Afaik:
* make
  * requires at least kustomize and kind (ci-e2e.sh installs them via hack/ensure-*.sh)
* scripts/ci-e2e.sh
  *  when new kindest/node images have to be build (i.e. they don't exist remotely) it builds the images locally and depends on the existence of "$GOPATH/src/k8s.io/kubernetes"
* IDE: same as scripts/ci-e2e.sh, because it's used to build the kindest/node images

Looking at `ci-e2e.sh` we make sure to have the following installed: go, kubectl, kustomize, kind.

I can document this on a best effort basis. Maybe @fabriziopandini knows some additional dependencies.

Reviewer 2: Let's start with what we have now, and then we can eventually iterate adding more context.
However, generally speaking we should document in single place the dependencies for a developer workstation, and reference this list whenever necessary, like in this page.
As of today we already have something similar in two places, which is not ideal
- https://master.cluster-api.sigs.k8s.io/developer/guide.html
- https://master.cluster-api.sigs.k8s.io/developer/tilt.html

So I think we should defer fixing this to a follow up PR

Reviewer: > I think it depends on which exact test is executed and in which way it is executed. 

@sbueringer, i was just thinking about the bare minimum, not necessarily all the options

> As of today we already have something similar in two places, which is not ideal

++, would be nice to have a single place that we could link

`make docker-build-e2e` will build the images for all providers that will be needed for the e2e tests.

### Test execution via ci-e2e.sh

> To run the tests locally via command line we can just take a look at the Job configuration and execute the same locally.

Reviewer: since we are instructing users on how they can any test for themselves, maybe make this next sentence a little more direct. 
```suggestion
To run a test locally via the command line, you should look at the Prow Job configuration for the test you want to run and then execute the same container commands locally.
```

For example to run [pull-cluster-api-e2e-main](https://github.com/kubernetes/test-infra/blob/49ab08a6a2a17377d52a11212e6f1104c3e87bfc/config/jobs/kubernetes-sigs/cluster-api/cluster-api-presubmits-main.yaml#L113-L140)
just execute:

```bash
GINKGO_FOCUS="\[PR-Blocking\]" ./scripts/ci-e2e.sh
```

### Test execution via make test-e2e

`make test-e2e` will run e2e tests by using whatever provider images already exist on disk.

> After running `make docker-build-e2e` at least once, this can be used for a faster test run, if there are no provider code changes.

Reviewer: minor nit
```suggestion
After running `make docker-build-e2e` at least once, `make test-e2e` can be used for a faster test run if there are no provider code changes.
```

Reviewer: also, what if there are provider code changes?

maybe add a note about recompiling if there are changes?

Author: Will do
